packages/contentful--react-apps-toolkit/src/SDKProvider.tsx

```diff
@@ -7,15 +7,35 @@ interface SDKProviderProps {
   loading?: ReactElement;
 }
 
+const DELAY_TIMEOUT = 4 * 1000;
+
 /**
  * The Component providing the AppSdk, the useSDK hook can only be used within this Provider
  * @param props.loading an optional loading element that gets rendered while initializing the app
  */
 export const SDKProvider: FC<SDKProviderProps> = (props) => {
   const [sdk, setSDK] = React.useState<KnownSDK | undefined>();
+  const timeout = useRef<number | undefined>();
+
+  const clearPossibleTimeout = () => {
+    if (timeout.current !== undefined) {
+      window.clearTimeout(timeout.current);
+    }
+  };
+
+  const onInitSuccess = (sdk: KnownSDK) => {
+    setSDK(sdk);
+    clearPossibleTimeout();
+  };
 
   React.useEffect(() => {
-    init(setSDK);
+    timeout.current = window.setTimeout(() => {
+      console.warn(
+        'Your app is taking longer than expected to initialize. If this is not fixed, let us know: https://github.com/contentful/ui-extensions-sdk/issues'
+      );
+    }, DELAY_TIMEOUT);
+    init(onInitSuccess);
+    return () => clearPossibleTimeout();
```

Comment:
Your solution feels quite verbose to me. Can't we just do this and remove the outside methods and imports? There is no need to store the timeout on component scope.

```suggestion
    const timeout = window.setTimeout(() => {
      console.warn(
        'Your app is taking longer than expected to initialize. If this is not fixed, let us know: https://github.com/contentful/ui-extensions-sdk/issues'
      );
    }, DELAY_TIMEOUT);
    init((sdk) => {
      setSDK(sdk);
      clearTimeout(timeout);
    });
    return () => clearTimeout(timeout);
```

Reply:
Wouldn't you get the `console.warn` in case of a re-render where one of the new renders is successful?

Reply:
You are right andi, since we don't use the timeout outside of the useEffect we can just keep it there👍